kernel/src/hil/kv_store.rs

```diff
@@ -0,0 +1,81 @@
+//! Interface for KV Stores
+
+use crate::returncode::ReturnCode;
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum Operation {
+    None,
+    Init,
+    GetKey,
+    AppendKey,
+    InvalidateKey,
+    GarbageCollect,
+}
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// This callback is called when any operation is complete.
+    ///
+    /// 'operation': The operation that completed
+    /// 'result': Nothing on success, 'ReturnCode' on error
+    /// 'key': A key buffer is one exists
+    /// 'value': A value buffer is one exists
+    /// 'buf': A buf buffer is one exists
+    fn operation_complete(
+        &'a self,
+        operation: Operation,
+        result: Result<(), ReturnCode>,
+        key: Option<&'static mut [u8]>,
+        value: Option<&'static mut [u8]>,
+        buf: Option<&'static mut [u8]>,
+    );
+}
+
+pub trait KVStore {
+    /// Appends the key/value pair.
+    ///
+    /// `key`: A unhashed key. This will be hashed internally. This key
+    ///        will be used in future to retrieve or remove the `value`.
+    /// `value`: A buffer containing the data to be stored to flash.
+    ///
+    /// On success nothing will be returned.
+    /// On error the key, value and a `ReturnCode` will be returned.
+    fn append_key(
+        &self,
+        key: &'static [u8],
+        value: &'static [u8],
+    ) -> Result<(), (&'static [u8], &'static [u8], ReturnCode)>;
+
+    /// Retrieves the value from a specified key.
+    ///
+    /// `key`: A unhashed key. This will be hashed internally.
+    /// `buf`: A buffer to store the value to.
+    ///
+    /// On success nothing will be returned.
+    /// On error the key, buf and a `ReturnCode` will be returned.
+    ///
+    /// If a power loss occurs before success is returned the data is
+    /// assumed to be lost.
+    fn get_key(
+        &self,
+        key: &'static [u8],
+        buf: &'static mut [u8],
+    ) -> Result<(), (&'static [u8], &'static [u8], ReturnCode)>;
+
+    /// Invalidates the key in flash storage
+    ///
+    /// `key`: A unhashed key. This will be hashed internally.
+    ///
+    /// On success nothing will be returned.
+    /// On error the key and a `ReturnCode` will be returned.
+    ///
+    /// If a power loss occurs before success is returned the data is
+    /// assumed to be lost.
+    fn invalidate_key(&self, key: &'static [u8]) -> Result<(), (&'static [u8], ReturnCode)>;
+
+    /// Perform a garbage collection on the KV Store
```

Comment:
Maybe it's worth adding a comment that on a key-value store which does not require garbage collection, this can be implemented as a `nop` and returning `Ok(0)`?